contracts/voting_escrow_delegation/tests/simulation.rs

```diff
@@ -0,0 +1,399 @@
+
+struct Simulator {
+    helper: Helper,
+    router: App,
+}
+
+impl Simulator {
+    fn new<T: Clone + Into<String>>(_users: &[T]) -> Self {
+        let mut router = mock_app();
+        Self {
+            helper: Helper::init(&mut router, Addr::unchecked("owner")),
+            router,
+        }
+    }
+
+    fn mint(&mut self, user: &str, amount: u128) {
+        self.helper
+            .escrow_helper
+            .mint_xastro(&mut self.router, user, amount as u64)
+    }
+
+    fn create_lock(&mut self, user: &str, amount: f64, interval: u64) -> Result<AppResponse> {
+        self.helper
+            .escrow_helper
+            .create_lock(&mut self.router, user, interval, amount as f32)
+    }
+
+    fn app_next_period(&mut self) {
+        self.router.update_block(next_block);
+        self.router
+            .update_block(|block| block.time = block.time.plus_seconds(WEEK));
+    }
+
+    fn create_delegation(
+        &mut self,
+        user: &str,
+        percentage: Uint128,
+        expire_time: u64,
+        token_id: String,
+        recipient: String,
+    ) -> Result<AppResponse> {
+        self.helper.create_delegation(
+            &mut self.router,
+            user,
+            percentage,
+            expire_time,
+            token_id,
+            recipient,
+        )
+    }
+
+    fn extend_delegation(
+        &mut self,
+        user: &str,
+        percentage: Uint128,
+        expire_time: u64,
+        token_id: String,
+    ) -> Result<AppResponse> {
+        self.helper
+            .extend_delegation(&mut self.router, user, percentage, expire_time, token_id)
+    }
+
+    fn event_router(&mut self, user: &str, event: Event) {
+        println!("User {} Event {:?}", user, event);
+        match event {
+            Event::CreateDelegation(percentage, expire_time, token_id, recipient) => {
+                if let Err(err) =
+                    self.create_delegation(user, percentage, expire_time, token_id, recipient)
+                {
+                    dbg!(err);
+                }
+            }
+            Event::ExtendDelegation(percentage, expire_time, token_id) => {
+                if let Err(err) = self.extend_delegation(user, percentage, expire_time, token_id) {
+                    dbg!(err);
+                }
+            }
+        }
+    }
+}
+
+use proptest::prelude::*;
+
+const MAX_PERIOD: usize = 10;
+const MAX_USERS: usize = 6;
+const MAX_TOKENS: usize = 10;
+const MAX_EVENTS: usize = 100;
+
+fn events_strategy() -> impl Strategy<Value = Event> {
+    prop_oneof![
+        (
+            1u64..=100u64,
+            1..MAX_PERIOD,
+            prop::collection::vec("[a-z]{4,32}", 1..MAX_USERS),
+            prop::collection::vec("[t-z]{6,32}", 1..MAX_TOKENS)
+        )
+            .prop_map(|(a, b, c, d)| {
+                Event::CreateDelegation(
+                    Uint128::from(a),
+                    WEEK * b as u64,
+                    c.iter().next().unwrap().to_string(),
```

Comment:
You need to select user from the line 136. Do not generate new users set in this strategy.

Reply:
Why ? We need to generate recipients. On the line 136 we generate users which will creates delegations

Reply:
Ah, these are delegates. Got it